src/main/java/tds/exam/results/mappers/OpportunityMapper.java

```diff
@@ -55,75 +69,82 @@
-                ? ADMINISTRATION_CONDITION_INVALIDATED
-                : "");
+        opportunity.setWindowOpportunity(String.valueOf(expandableExam.getWindowAttempts()));
+        opportunity.setAdministrationCondition(exam.getStatus().getCode().equalsIgnoreCase(ExamStatusCode.STATUS_INVALIDATED)
+            ? ADMINISTRATION_CONDITION_INVALIDATED
+            :  StringUtils.EMPTY);
         opportunity.setDateForceCompleted(expandableExam.getForceCompletedAt() != null
-            ? expandableExam.getForceCompletedAt().toString()
-            : null); //TODO: Verify date format
+            ? expandableExam.getForceCompletedAt().toDateTime().toString()
+            : null);
         opportunity.setClientName(exam.getClientName());
         opportunity.setAssessmentParticipantSessionPlatformUserAgent(exam.getBrowserUserAgent());
-//        opportunity.setEffectiveDate(); TODO: Add effectiveDate to ExpandableExam
+        opportunity.setEffectiveDate(!assessmentWindows.isEmpty()
+            ? assessmentWindows.get(0).getStartTime().toString() //TODO: Verify date format
+            : StringUtils.EMPTY);
 
-        mapExamSegmentsToOpportunity(expandableExam.getExamSegments(), opportunity.getSegment());
+        mapExamSegmentsToOpportunity(expandableExam.getExamSegmentWrappers(), opportunity.getSegment());
         mapExamAccommodationsToOpportunity(expandableExam.getExamAccommodations(),  opportunity.getAccommodation());
 
         // Map all of the assessment items with the item id as the map's key
         final Map<String, Item> assessmentItems = assessment.getSegments().stream()
             .flatMap(segment -> segment.getItems().stream())
             .collect(Collectors.toMap(Item::getId, Function.identity(), (item1, item2) -> item1));
 
-        mapExamItemsAndResponsesToOpportunity(expandableExam, opportunity.getItem(), assessmentItems);
+        mapExamItemsAndResponsesToOpportunity(expandableExam, examItems, opportunity.getItem(), assessmentItems);
 
         return opportunity;
     }
 
     private static void mapExamItemsAndResponsesToOpportunity(final ExpandableExam expandableExam,
+                                                              final List<ExamItem> examItems,
                                                               final List<TDSReport.Opportunity.Item> opportunityItems,
                                                               final Map<String, Item> assessmentItems) {
-        for (ExamItem examItem : expandableExam.getExamItems()) {
-            // find the page for this exam item - this should never be "empty"
-            final ExamPage examPage = expandableExam.getExamPages().stream()
-                .filter(page -> page.getId().equals(examItem.getExamPageId()))
+        Map<UUID, Integer> itemVisitsMap = expandableExam.getItemResponseUpdates();
+        final List<ExamSegment> examSegments = expandableExam.getExamSegmentWrappers().stream()
```

Comment:
What about collecting this to a `Map<String,ExamSegment>` where the String is the segment key?  Then you can get rid of `getSegmentIdFromSegmentKey` and instead just do a key lookup on line 119.  It doesn't seem you ever need a List of ExamSegments anywhere else.